alf/environments/parallel_environment.cpp

```diff
@@ -0,0 +1,501 @@
+                                       const std::string& name)
+    : EnvironmentBase(num_envs, action_spec, timestep_spec, name),
+      name_(name),
+      env_id_(env_id),
+      env_(env),
+      call_handler_(call_handler),
+      py_step_(env.attr("step")),
+      py_reset_(env.attr("reset")),
+      job_queue_(bip::open_or_create,
+                 (name + ".job_queue." + std::to_string(env_id)).c_str(),
+                 3,
+                 sizeof(Job)),
+      ready_queue_(bip::open_or_create,
+                   (name + ".ready_queue").c_str(),
+                   num_envs,
+                   sizeof(int)),
+      just_reseted_(false) {
+  env_id_buf_ =
+      reinterpret_cast<int32_t*>(timestep_buffer_.GetBufByFieldName("env_id"));
+}
+
+ProcessEnvironment::~ProcessEnvironment() {
+  bip::message_queue::remove(
+      (name_ + ".job_queue." + std::to_string(env_id_)).c_str());
+  bip::message_queue::remove((name_ + ".ready_queue").c_str());
+}
+
+class ProcessEnvironmentCaller {
+  std::string name_;
+  int env_id_;
+  bip::message_queue job_queue_;
+
+ public:
+  ProcessEnvironmentCaller(int env_id, const std::string& name)
+      : name_(name),
+        env_id_(env_id),
+        job_queue_(bip::open_or_create,
+                   (name + ".job_queue." + std::to_string(env_id)).c_str(),
+                   3,
+                   sizeof(Job)) {}
+  ~ProcessEnvironmentCaller() {
+    bip::message_queue::remove(
+        (name_ + ".job_queue." + std::to_string(env_id_)).c_str());
+  }
+
+  void Call();
+  void Close();
+};
+
+void ProcessEnvironmentCaller::Call() {
+  Job job{JobType::call};
+  job_queue_.send(&job, sizeof(job), 0);
+}
+
+void ProcessEnvironmentCaller::Close() {
+  Job job{JobType::close};
+  job_queue_.send(&job, sizeof(job), 0);
+}
+
+void ProcessEnvironment::Step(int env_id) {
+  py::object timestep;
+  if (just_reseted_) {
+    timestep = reset_result_;
+    just_reseted_ = false;
+  } else {
+    auto action = action_buffer_.ViewAsNestedArray(env_id);
+    timestep = py_step_(action);
+  }
+  SendTimestep(timestep, env_id);
+}
+
+void ProcessEnvironment::Reset() {
+  reset_result_ = py_reset_();
+  just_reseted_ = true;
+}
+
+void ProcessEnvironment::SendTimestep(const py::object& timestep, int env_id) {
+  timestep_buffer_.Write(timestep, env_id);
+  env_id_buf_[env_id] = env_id;
+  ready_queue_.send(&env_id, sizeof(env_id), 0);
+}
+
+void ProcessEnvironment::Worker() {
+  while (true) {
+    Job job;
+    bip::message_queue::size_type recvd_size;
+    unsigned int priority;
+    job_queue_.receive(&job, sizeof(job), recvd_size, priority);
+    assert(recvd_size == sizeof(job));
+    if (job.type == JobType::step) {
+      Step(job.env_id);
+    } else if (job.type == JobType::reset) {
+      Reset();
+    } else if (job.type == JobType::close) {
+      env_.attr("close")();
+      break;
+    } else if (job.type == JobType::call) {
+      call_handler_();
```

Comment:
I am still a bit confused by this. It seems that `call_handler_()` is just the original python `ProcessEnvironment`'s handler. Why do we still need it? This has a connection with a follow-up question below.